File: 백준/Platinum/13511. 트리와 쿼리 2/트리와 쿼리 2.py
# 트리와 쿼리2
import sys
input = sys.stdin.readline
import math
from collections import deque

# depth 계산
def calDepth(n):
    que = deque([n])
    depth[n] = 1

    while que:
        cur = que.popleft()

        for cn, dist in graph[cur]:
            if dp[cur][0][0] != cn:
                dp[cn][0] = (cur, dist)
                depth[cn] = depth[cur] + 1
                que.append(cn)

# n번 노드 기준 dist 위에 존재하는 부모 노드 검색
def findNode(n, dist):
    result = n
    for i in range(LOG, -1, -1):
        if dist >= 1 << i:
            dist -= 1 << i
            result = dp[result][i][0]
        if dist == 0:
            break
        
    return result

# LCA를 계산해 총 이동거리, LCA, 각 노드별 LCA까지 이동횟수 리턴
def LCA(n1, n2):
    total = 0 # 총 거리 
    dist_n1 = 0 # LCA가 n1의 몇번째 부모인지 담는 변수
    dist_n2 = 0 # LCA가 n2의 몇번째 부모인지 담는 변수
    isChange = False
    
    if depth[n1] > depth[n2]:
        n1, n2 = n2, n1
        isChange = True
        
    for i in range(LOG, -1, -1):
        if depth[n2] - depth[n1] >= 1 << i:
            total += dp[n2][i][1]
            dist_n2 += 1 << i
            n2 = dp[n2][i][0]

    if isChange:
        n1, n2 = n2, n1
        dist_n1, dist_n2 = dist_n2, dist_n1
    
    if n1 == n2:
        return (total, dist_n1, dist_n2)

    for i in range(LOG, -1, -1):
        if dp[n1][i][0] != dp[n2][i][0]:
            total += dp[n1][i][1]
            total += dp[n2][i][1]
            dist_n1 += 1 << i
            dist_n2 += 1 << i
            n1 = dp[n1][i][0]
            n2 = dp[n2][i][0]
            
    total += dp[n1][0][1]
    total += dp[n2][0][1]
    return (total, dist_n1+1, dist_n2+1)

n = int(input())
graph = [[] for _ in range(n+1)]
LOG = int(math.log(n, 2))
depth = [0] * (n+1)
# dp[i][j] = i의 2^j번째 (조상노드, 거리)
dp = [[(0, 0)] * (LOG+1) for _ in range(n+1)]

for i in range(n-1):
    n1, n2, dist = map(int,input().split())
    graph[n1].append((n2, dist))
    graph[n2].append((n1, dist))

calDepth(1)
for i in range(1, LOG+1):
    for j in range(1, n+1):
        prev = dp[j][i-1][0]
        dist = dp[j][i-1][1] + dp[prev][i-1][1]
        dp[j][i] = (dp[prev][i-1][0], dist)
        
m = int(input())
for i in range(m):
    cmd = list(map(int,input().split()))
    u = cmd[1]
    k = cmd[2]
    total, dist_n1, dist_n2 = LCA(u, k)

    if cmd[0] == 1:
        print(total)
        continue
    
    seq = cmd[3]
    
   
    if dist_n1 >= seq: # 찾고자하는 노드가 n1노드부터 LCA 사이에 있는 경우
        print(findNode(u, seq - 1))
    else: # 찾고자하는 노드가 n2노드부터 LCA 사이에 있는 경우
        print(findNode(k, dist_n2 - (seq - dist_n1 - 1)))
    

File: 백준/Platinum/13511. 트리와 쿼리 2/트리와 쿼리 2.java
import java.io.*;
import java.util.*;

public class Main {
    static int depth[];
    static long[][][] dp;
    static ArrayList<int[]>[] graph;
    static int LOG;

    static void calDepth(int n) {
        ArrayDeque<Integer> que = new ArrayDeque<>();
        que.addLast(n);
        depth[n] = 1;

        while (!que.isEmpty()) {
            int cur = que.removeFirst();

            for (int[] i : graph[cur]) {
                int child = i[0];
                int distance = i[1];
                if (dp[cur][0][0] == child) {
                    continue;
                }
                dp[child][0][0] = cur;
                dp[child][0][1] = distance;
                depth[child] = depth[cur] + 1;
                que.addLast(child);
            }
        }
    }

    static long findNode(int n, long dist) {
        int result = n;
        for (int i = LOG; i >= 0; i--) {
            if (dist >= 1 << i) {
                result = (int) dp[result][i][0];
                dist -= 1 << i;
            }

            if (dist == 0) {
                break;
            }
        }
        return result;
    }

    public static long[] LCA(int n1, int n2) {
        boolean isChange = false;
        long distN1 = 0;
        long distN2 = 0;
        long totalDist = 0;

        if (depth[n1] > depth[n2]) {
            isChange = true;
            int tmp = n1;
            n1 = n2;
            n2 = tmp;
        }

        for (int i = LOG; i >= 0; i--) {
            if (depth[n2] - depth[n1] >= 1 << i) {
                totalDist += dp[n2][i][1];
                distN2 += 1 << i;
                n2 = (int) dp[n2][i][0];
            }
        }

        if (isChange) {
            int tmp = n1;
            n1 = n2;
            n2 = tmp;
            long distTmp = distN1;
            distN1 = distN2;
            distN2 = distTmp;
        }

        if (n1 == n2) {
            return new long[] {totalDist, n1, distN1, distN2};
        }

        for (int i = LOG; i >= 0; i--) {
            if (dp[n1][i][0] != dp[n2][i][0]) {
                totalDist += dp[n1][i][1];
                totalDist += dp[n2][i][1];
                distN1 += 1 << i;
                distN2 += 1 << i;
                n1 = (int) dp[n1][i][0];
                n2 = (int) dp[n2][i][0];
            }
        }

        totalDist += dp[n1][0][1];
        totalDist += dp[n2][0][1];
        return new long[] {totalDist, dp[n1][0][0], distN1+1, distN2+1};
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int n = Integer.parseInt(br.readLine());
        LOG = (int) (Math.log(n) / Math.log(2));
        dp = new long [n+1][LOG+1][2];
        depth = new int[n+1];
        graph = new ArrayList[n+1];
        for (int i = 0; i < n+1; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < n-1; i++) {
            st = new StringTokenizer(br.readLine());
            int n1 = Integer.parseInt(st.nextToken());
            int n2 = Integer.parseInt(st.nextToken());
            int dist = Integer.parseInt(st.nextToken());

            graph[n1].add(new int[] {n2, dist});
            graph[n2].add(new int[] {n1, dist});
        }

        calDepth(1);
        for (int i = 1; i <= LOG; i++) {
            for (int j = 1; j <= n; j++) {
                int prev = (int) dp[j][i-1][0];
                dp[j][i][0] = dp[prev][i-1][0];
                dp[j][i][1] = dp[j][i-1][1] + dp[prev][i-1][1];
            }
        }

        int m = Integer.parseInt(br.readLine());
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int cmd = Integer.parseInt(st.nextToken());
            int n1 = Integer.parseInt(st.nextToken());
            int n2 = Integer.parseInt(st.nextToken());
            long[] result = LCA(n1, n2);
            long total = result[0];
            long parent = result[1];
            long distN1 = result[2];
            long distN2 = result[3];

            if (cmd == 1) {
                System.out.println(total);
            } else {
               int seq = Integer.parseInt(st.nextToken());
               if (seq == distN1+1) {
                   System.out.println(parent);
               } else if (distN1 >= seq) {
                   System.out.println(findNode(n1, seq - 1));
               } else {
                   System.out.println(findNode(n2, (distN2 - (seq - distN1 - 1))));
               }
            }
        }
    }
}
